### ISSUE TITLE
Incorrect Validation Accuracy Due to Distributed Sampler

### ISSUE THREAD

**Issue by SaltedFishLZ** [pytorch/pytorch#25162]
## 🐛 Bug

The new `DistributedSampler` pads the number of samples to make it divisible by the number of processes, which is a bit similar to `drop_last` in `DataLoader` while here `DsitributedSampler` duplicates the last few samples instead of dropping them.

This will definitely cause incorrect validation accuracy when you want to use distributed validation.

```Python
    def __iter__(self):
        # deterministically shuffle based on epoch
        g = torch.Generator()
        g.manual_seed(self.epoch)
        if self.shuffle:
            indices = torch.randperm(len(self.dataset), generator=g).tolist()
        else:
            indices = list(range(len(self.dataset)))


        # add extra samples to make it evenly divisible
        indices += indices[:(self.total_size - len(indices))]
        assert len(indices) == self.total_size

        # subsample
        indices = indices[self.rank:self.total_size:self.num_replicas]
        assert len(indices) == self.num_samples

        return iter(indices)
```

## Environment
 - PyTorch Version (1.2.0):

## Feature Request
Add `drop_last` or `duplicate_last` option in `DistributedSampler`.

cc @SsnL @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera

**Comment by pietern** [comment]
Good point. Is this something you're willing to contribute?

**Comment by mrcslws** [comment]
I made a similar fix. Note that lockup issues can theoretically occur if you remove the padding, since the DistributedDataParellel assumes each process is going to call forward an equal number of times. I describe this caveat in the docstring: https://github.com/numenta/nupic.research/blob/a2bf34227229f4b881fba1f480d2c4867a3cec20/nupic/research/frameworks/pytorch/distributed_sampler.py#L29-L37

**Comment by rohan-varma** [comment]
Another use case for supporting uneven inputs as described in https://github.com/pytorch/pytorch/issues/38174 - it would be useful to avoid the lockup issue @mrcslws described.

**Comment by rohan-varma** [comment]
This should be resolved by https://github.com/pytorch/pytorch/pull/41171. If you pass in `drop_last=True` into the constructor, the tail of the data will be dropped and there won't be duplicates. We still ensure that the sampling is even across all the processes.

**Comment by mrcslws** [comment]
Hmm... I don't think that solves the issue. We want correct validation accuracy. This fix doesn't provide correct validation accuracy, it just makes it incorrect in a different way. It discards part of the dataset rather than repeating part of it.

**Comment by SsnL** [comment]
I agree with @mrcslws . That is not a fix. Reopening.

**Comment by SsnL** [comment]
What we need really is just an option to return all samples, in possibly uneven batches. In validation phase we don’t care about balancing across ranks anyways since there is no gradient averaging. Users can sum gather the # correct predictions, and then divide by dataset size to obtain correct Val acc.

**Comment by rohan-varma** [comment]
Good point re: returning all samples (possibly uneven), related issue for that is https://github.com/pytorch/pytorch/issues/49180

**Comment by josh-gleason** [comment]
I really hope this issue will be addressed. This particular issue means that if you use DDP for validation or test then the number of gpus you use can impact your validation or test metrics, which is weird. Either I need to go remove duplicates afterwards or write my own sampler.

Turns out that with the 1.12 [implementation of DistributedDataSampler](https://github.com/pytorch/pytorch/blob/v1.12.1/torch/utils/data/distributed.py#L13) only a small change gives the desired results. Just keep in mind that this may cause the dataloader to return uneven batches or potentially a different number of batches on different processes.

```python
import torch.utils.data

class DistributedSamplerNoDuplicate(torch.utils.data.DistributedSampler):
    """ A distributed sampler that doesn't add duplicates. Arguments are the same as DistributedSampler """
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        if not self.drop_last and len(self.dataset) % self.num_replicas != 0:
            # some ranks may have less samples, that's fine
            if rank >= len(self.dataset) % self.num_replicas:
                self.num_samples -= 1
            self.total_size = len(self.dataset)
```

**Comment by MattGPT-ai** [comment]
Bumping this issue because I just came across this. Wondering if this is still a problem to implement, or the above-mentioned code works if you just replace `rank` -> `torch.distributed.get_rank()`
